app/src/main/java/org/mozilla/focus/browser/CursorViewModel.kt

```diff
@@ -0,0 +1,159 @@
+            field = value
+            clampPos(pos, value)
+        }
+
+    private val pos = PointF(0f, 0f)
+    private var vel = 0f
+
+    private val pressedDirections = mutableSetOf<Direction>()
+
+    private var updateLoop: Deferred<Unit>? = null
+
+    // TODO: Distances should update as a function of time to prevent lag.
+    // TODO: Describe algorithm more naturally as velocity and acceleration vectors.
+    private fun update(deltaMillis: Long) {
+        vel = (vel + 1) * FRICTION
+        vel = Math.min(MAX_SPEED, vel)
+        val isMovingDiagonal = pressedDirections.size > 1
+        val finalVel = if (isMovingDiagonal) vel / 2 else vel
+
+        for (dir in pressedDirections) {
+            updatePosForVel(dir, finalVel)
+        }
+
+        clampPos(pos, maxBounds)
+        onUpdate(pos.x, pos.y)
+    }
+
+    private fun updatePosForVel(dir: Direction, vel: Float) {
+        when (dir) {
+            Direction.UP -> pos.y -= vel
+            Direction.DOWN -> pos.y += vel
+            Direction.LEFT -> pos.x -= vel
+            Direction.RIGHT -> pos.x += vel
+        }
+    }
+
+    // TODO: stop when new views (e.g. overlay) are opened.
+    private fun asyncStartUpdates() = async(UI) { // Use UI to avoid synchronization.
+        var currentFrameMillis = SystemClock.uptimeMillis() // duped in loop.
+        var prevFrameMillis: Long
+        var deltaMillis = UPDATE_DELAY_MILLIS // Move ~1 frame to start.
+        while (isActive) {
+            update(deltaMillis)
+
+            delay(UPDATE_DELAY_MILLIS, TimeUnit.MILLISECONDS)
+
+            prevFrameMillis = currentFrameMillis
+            currentFrameMillis = SystemClock.uptimeMillis() // duped in init.
+            deltaMillis = currentFrameMillis - prevFrameMillis
+        }
+    }
+
+    /**
+     * Converts key events into Cursor actions; an analog to [Activity.dispatchKeyEvent].
+     *
+     * @return true if this key event was handled, false otherwise.
+     */
+    @UiThread
+    fun dispatchKeyEvent(event: KeyEvent): Boolean {
+        if (event.action != KeyEvent.ACTION_DOWN
+                && event.action != KeyEvent.ACTION_UP) return false
+
+        val remoteKey = RemoteKey.fromKeyEvent(event) ?: return false
+        if (remoteKey == RemoteKey.CENTER) {
+            dispatchTouchEventOnCurrentPosition(event.action)
+            return true
+        }
+
+        val direction = remoteKey.toDirection()
+        if (direction != null) {
+            onDirectionKey(direction, event.action)
+            return true
+        }
+        return false
+    }
+
+    private fun onDirectionKey(dir: Direction, action: Int) {
+        if (action == KeyEvent.ACTION_DOWN) {
+            pressedDirections.add(dir)
+            if (updateLoop == null) updateLoop = asyncStartUpdates()
+        } else if (action == KeyEvent.ACTION_UP) {
+            pressedDirections.remove(dir)
+            if (pressedDirections.isEmpty()) {
+                updateLoop?.cancel()
```

Comment:
If someone clicks these buttons really quickly, could `updateLoop` get updated in the wrong order and have its cancelled/created changes to the variable be interleaved? e.g., should it be volatile? I'm not sure how the async coroutine gets updated/cancelled. (e.g., if this could resurface something like #208)

Reply:
No – dispatchKeyEvent can only get called from the UI thread so the key events can only get processed in order (assuming the system sends them in order) and there's no need for volatile.

The async coroutine also runs on the main thread so no locking is necessary either.